orchagent/dash/dashaclgroupmgr.cpp

```diff
@@ -0,0 +1,721 @@
+            {{DashAclDirection::OUT, SAI_IP_ADDR_FAMILY_IPV4, DashAclStage::STAGE2}, SAI_ENI_ATTR_OUTBOUND_V4_STAGE2_DASH_ACL_GROUP_ID},
+            {{DashAclDirection::OUT, SAI_IP_ADDR_FAMILY_IPV4, DashAclStage::STAGE3}, SAI_ENI_ATTR_OUTBOUND_V4_STAGE3_DASH_ACL_GROUP_ID},
+            {{DashAclDirection::OUT, SAI_IP_ADDR_FAMILY_IPV4, DashAclStage::STAGE4}, SAI_ENI_ATTR_OUTBOUND_V4_STAGE4_DASH_ACL_GROUP_ID},
+            {{DashAclDirection::OUT, SAI_IP_ADDR_FAMILY_IPV4, DashAclStage::STAGE5}, SAI_ENI_ATTR_OUTBOUND_V4_STAGE5_DASH_ACL_GROUP_ID},
+            {{DashAclDirection::OUT, SAI_IP_ADDR_FAMILY_IPV6, DashAclStage::STAGE1}, SAI_ENI_ATTR_OUTBOUND_V6_STAGE1_DASH_ACL_GROUP_ID},
+            {{DashAclDirection::OUT, SAI_IP_ADDR_FAMILY_IPV6, DashAclStage::STAGE2}, SAI_ENI_ATTR_OUTBOUND_V6_STAGE2_DASH_ACL_GROUP_ID},
+            {{DashAclDirection::OUT, SAI_IP_ADDR_FAMILY_IPV6, DashAclStage::STAGE3}, SAI_ENI_ATTR_OUTBOUND_V6_STAGE3_DASH_ACL_GROUP_ID},
+            {{DashAclDirection::OUT, SAI_IP_ADDR_FAMILY_IPV6, DashAclStage::STAGE4}, SAI_ENI_ATTR_OUTBOUND_V6_STAGE4_DASH_ACL_GROUP_ID},
+            {{DashAclDirection::OUT, SAI_IP_ADDR_FAMILY_IPV6, DashAclStage::STAGE5}, SAI_ENI_ATTR_OUTBOUND_V6_STAGE5_DASH_ACL_GROUP_ID},
+        };
+
+    auto stage = StageMaps.find({d, f, s});
+    if (stage == StageMaps.end())
+    {
+        SWSS_LOG_ERROR("Invalid stage %d %d %d", static_cast<int>(d), f, static_cast<int>(s));
+        throw runtime_error("Invalid stage");
+    }
+
+    return stage->second;
+}
+
+DashAclGroupMgr::DashAclGroupMgr(DashOrch *dashorch, DashAclOrch *aclorch) :
+    m_dash_orch(dashorch),
+    m_dash_acl_orch(aclorch)
+{
+    SWSS_LOG_ENTER();
+}
+
+void DashAclGroupMgr::init(DashAclGroup& group)
+{
+    SWSS_LOG_ENTER();
+    group.m_dash_acl_group_id = SAI_NULL_OBJECT_ID;
+
+    for (auto& rule: group.m_dash_acl_rule_table)
+    {
+        rule.second.m_dash_acl_rule_id = SAI_NULL_OBJECT_ID;
+    }
+}
+
+sai_status_t DashAclGroupMgr::create(DashAclGroup& group)
+{
+    SWSS_LOG_ENTER();
+
+    vector<sai_attribute_t> attrs;
+
+    attrs.emplace_back();
+    attrs.back().id = SAI_DASH_ACL_GROUP_ATTR_IP_ADDR_FAMILY;
+    attrs.back().value.s32 = group.m_ip_version;
+
+    auto status = sai_dash_acl_api->create_dash_acl_group(&group.m_dash_acl_group_id, gSwitchId, static_cast<uint32_t>(attrs.size()), attrs.data());
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        return status;
+    }
+
+    CrmResourceType crm_rtype = (group.m_ip_version == SAI_IP_ADDR_FAMILY_IPV4) ?
+        CrmResourceType::CRM_DASH_IPV4_ACL_GROUP : CrmResourceType::CRM_DASH_IPV6_ACL_GROUP;
+    gCrmOrch->incCrmDashAclUsedCounter(crm_rtype, group.m_dash_acl_group_id);
+
+    return status;
+}
+
+task_process_status DashAclGroupMgr::create(const string& group_id, DashAclGroup& group)
+{
+    SWSS_LOG_ENTER();
+
+    if (exists(group_id))
```

Comment:
Is there a possibility to get a request to create a group twice?

Reply:
We can check the group attributes and if they are the same return and warning instead of error. But I'm not sure that this can happen in the production. If duplicates are not possible the is no reason to additional logic to the code.